content/blog/2025-09-15-hash-fields-expiration/index.md

```diff
@@ -0,0 +1,225 @@
+Why not always use a very small time window (e.g few milliseconds) for each bucket? The reason is memory.
+Recall that the buckets are managed in a Radix-Tree which introduces a high overhead per each item.
+When many buckets are used, the memory overhead can be high and we might end up with the same memory overhead introduced by using a Radix-Tree index.
+
+The 'vset' also uses different bucket encoding based on the number of items in the bucket.
+A bucket with only a single element would require just a single pointer size of bytes.
+A bucket of small amount of items would be encoded as a vector of item pointers, and when the bucket contains many items we will use the [Valkey hashtable](/blog/new-hash-table) structure to map the relevant items.
+By leveraging data structures we already know how to optimize, we can further improve memory usage and performance with techniques like SIMD acceleration and memory prefetching.
+
+A hash object that contains volatile fields now also carries a secondary vset index.
+At the database level, we maintain a global map of hashes with volatile fields.
+The active expiration cron job scans both regular keys and these hashes, but only iterates over vset buckets whose end time has passed.
+This ensures that CPU time is spent only on fields that are truly ready to expire.
+
+On the storage side, expiration metadata is stored compactly, inlined alongside the hash’s key-value entries.
+If an expired field is accessed before the cron job has removed it, Valkey treats it as if it doesn’t exist (with a few well-documented exceptions).
+
+### Known limitations
+
+To deliver hash field expirations in a way that is simple, performant, and memory-efficient, we made some deliberate trade-offs in the initial implementation.
+As a result, there are a few limitations today that we plan to revisit in future releases:
+
+1. Expired fields are treated as if they don’t exist when accessed through commands like [`HGET`](/commands/hget/) or [`HGETALL`](/commands/hgetall/).
+   However, [`HLEN`](/commands/hlen/) provides an exception for this logic.
+   Because [`HLEN`](/commands/hlen/) is optimized to read only the metadata tracking the total number of fields in a hash, it may still count expired fields that haven’t yet been cleaned up by the background expiration job.
+   This means that until cleanup occurs, [`HLEN`](/commands/hlen/) might temporarily report a higher count than the number of “live” fields.
+2. The [`HRANDFIELD`](/commands/hrandfield/) command selects fields at random from the underlying hash object.
+   If a hash contains many expired fields awaiting deletion, the selection process may frequently land on those expired entries.
+   In practice, this can result in [`HRANDFIELD`](/commands/hrandfield/) returning an empty reply more often than expected, and it reduces the fairness of field sampling.
+3. Valkey normally uses a highly compact representation for small hashes, which saves significant memory.
```

Comment:
I wouldn't call this a limitation, I would say this is a next step and area for improvement.

Reply:
I think I will take your suggesion to just drop this section